workflow/controller/operator.go

```diff
@@ -1199,6 +1190,29 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatu
 	return nil
 }
 
+func getPodTemplate(pod *apiv1.Pod) (*wfv1.Template, error) {
+	tmpl := &wfv1.Template{}
+	for _, c := range pod.Spec.Containers {
+		for _, e := range c.Env {
+			if e.Name == common.EnvVarTemplate {
```

Comment:
I don’t think we should use this annotation here. Pod names are deterministic and the same as the node ID. So you can look that up in status.Nodes[pod.Name] and then user woc.wf.FindTemplate. This way we can avoid any accidental breakage when an old pod did not have the annotation.

Reply:
True. Actually the template is only used to get the `daemon` value, so I replaced the call with `woc.findTemplate()`. The `getPodTemplate()` function was moved to `operator_test.go`, since it is only used to verify the presence of variables in tests.